In [98]:
import pandas as pd
import numpy as np
import re

In [5]:
from konlpy.tag import Twitter
twitter = Twitter()

C:\Users\sbg02\Anaconda3\lib\site-packages\konlpy\tag\_okt.py:16: UserWarning: "Twitter" has changed to "Okt" since KoNLPy v0.4.5.
  warn('"Twitter" has changed to "Okt" since KoNLPy v0.4.5.')


In [105]:
df = pd.read_csv('C:/newsdata/삼성전자.csv',  encoding = 'utf-8 sig', engine = 'python')
df.head()

TITLE  \
0                  삼성전자 강경훈 부사장 19일 구속심사   
1      檢 '노조 와해 의혹' 강경훈 삼성전자 부사장 구속영장 청구   
2     "사회가 건강해야 기업도 발전" 20년 넘은 삼성 나눔의 비결   
3  4차 산업 혁신기업 인탑스, 스마트팩토리 중소벤처기업부 장관상 수상   
4        '에버랜드 노조와해' 삼성 노사총괄 부사장 내일 영장심사   

                                                BODY             TIME  
0  [theL]삼성에버랜드 노조 설립 방해·노조 와해 혐의에버랜드 노조 와해 의혹을 수...   2018.12.181953  
1  [the L]'삼성 노조 와해 의혹'을 수사 중인 검찰이 강경훈 삼성전자 부사장에 ...   2018.12.181943  
2  [특집]삼성전자는 1995년 국내 기업 최초로 사회봉사단을 창단 사회문제 해결을 위...   2018.12.181830  
3  인탑스(주)(대표 김근하 정사진)가 스마트팩토리 도입을 추진해 경북창조경제혁신센터가...  &2018.12.181745  
4  염호석씨 시신탈취 과정서 뇌물수수 경찰관도 함께(서울=뉴스1) 나연준 기자 = 삼성...  12018.12.181710

### TIME클렌징1 - TIME칼럼 EDA
1. for문 => 너무 복잡 & 시간이 필요<br>
<코드>  ???
2. RE => str으로 변환 뒤 '12018.' => '2018.'  으로 바꿔야 하는데 결과값이 Series가 아니라 \t처리된 str으로 나온다.<br>
<코드> pre3 = re.sub('[1\d]{4}8\.', '2018.', str(pre1))

### (아래에서 끌어온 결과)
#### 데이터 3등분 
<코드> time['year'].unique()

=> 결과<br>
array(['2018', '12018', '-2018', '', '--2018', '3002018', '22018',
       '+2018', '122018', '2017', '12017', '-2017', '22017', '112017',
       '122017']<br><br>
=> 뒤 숫자 4개만 남기면 정리될 듯?

### 연도 클렌징 - 특수문자 &클렌징    (뒤에 더 좋은 방법이 있어서 불필요할듯...)

In [141]:
pre1 = pd.Series(df['TIME']).str.replace('&', '')
pre1.head()

0     2018.12.181953
1     2018.12.181943
2     2018.12.181830
3     2018.12.181745
4    12018.12.181710
Name: TIME, dtype: object

### 연도 클렌징 - '.' 기준으로 칼럼 3등분 나누어 접근

In [171]:
time = pd.DataFrame(pre1.str.split('.',2).tolist(), columns = ['year','month','time'])
time.head()

year month    time
0   2018    12  181953
1   2018    12  181943
2   2018    12  181830
3   2018    12  181745
4  12018    12  181710

In [147]:
'''
https://www.geeksforgeeks.org/python-pandas-series-str-slice/
year_clean = time.year[-4:]

for i in range(0,len(df)):
    df.iloc[i].col = df.iloc[i].col[:9]
    
df.col = df.col.str[:9]
    
    
    '''

8276    2017
8277    2017
8278    2017
8279    2017
Name: year, dtype: object

### 연도 클렌징(실패) - 슬라이싱한 부분을 삭제하는 코드....ㅠ

In [156]:
'''
# start stop and step variables 
start, stop, step = 0, -4, 1
  
# converting to string data type 
data["Salary"]= data["Salary"].astype(str) 
  
# slicing till 4rd last element 
time['year'] = time['year'].str.slice(start, stop, step) 
  
# display 
time.head()
'''

year month    time
0         12  181953
1         12  181943
2         12  181830
3         12  181745
4    1    12  181710

### 연도 클렌징(성공) - 람다 사용해 단순 백슬라이싱

In [172]:
time['year'] = time['year'].apply(lambda x: x[-4:])
time['year'].unique()

array(['2018', '', '2017'], dtype=object)

### 시간 클렌징 - 초 삭제

In [173]:

# start stop and step variables 
start, stop, step = 0, -2, 1
  
# slicing till 2nd last element 
time['time'] = time['time'].str.slice(start, stop, step) 
  
# display 
time.head()

year month  time
0  2018    12  1819
1  2018    12  1819
2  2018    12  1818
3  2018    12  1817
4  2018    12  1817

### 마무리 - 원본에 붙이기

In [177]:
# 칼럼 삭제
del df['TIME']

KeyError: 'TIME'

In [180]:
# 칼럼 합치기
df_timeok = pd.concat([df, time], axis=1, sort=False)
df_timeok

TITLE  \
0                                 삼성전자 강경훈 부사장 19일 구속심사   
1                     檢 '노조 와해 의혹' 강경훈 삼성전자 부사장 구속영장 청구   
2                    "사회가 건강해야 기업도 발전" 20년 넘은 삼성 나눔의 비결   
3                 4차 산업 혁신기업 인탑스, 스마트팩토리 중소벤처기업부 장관상 수상   
4                       '에버랜드 노조와해' 삼성 노사총괄 부사장 내일 영장심사   
5                       文 "수소차에 획기적 뒷받침…비장한 각오로 제조업 부흥"   
6                             코스피, 외인·개인 매도에 2060선으로 밀려   
7                   삼성전자, 美 연말 초대형 TV 판매 웃었다…中·日 추격 따돌려   
8     [부자들은 경제신문을 읽는다] 김성훈 팀장 "테마주, 고점 매수No… 손절가 확인 ...   
9         [돈 잘버는 증권맨에게 듣는다] 이재광 이사 "아난티 이상 급등… 과감히 매수?"   
10                    文 "제조업 혁신, 노동자-기업 양보하며 상생모델 만들어야"   
11                             이스탄불 신공항 장식한 삼성 LED 사이니지   
12                        방 안에 옷가게…삼성 패션 "집에서 입어보고 사세요"   
13                            [특징주]삼성전자, 사흘째 52주 신저가 경신   
14                       美 뉴욕증시 하락에 코스피도 내림세...2060선 약세   
15                         [MT리포트] 中 화웨이는 왜 '공공의 적'이 됐나   
16                      [부고]신원빈씨(머니투데이 국제부 디자인기자) 부친상 외   
17                         [부고]박헌용씨(SK이노베이션 CR전략실장) 부친상   
18            [MT리포트]화웨이 韓 시장 공략 어디까지?…모바일·은행망·지하철까지 진출   
19                   [MT리포트]美 '중국 기술 굴기' 정조준, 中 화웨이 운명은   
20                                삼성전자 강경훈 부사장 19일 구속심사   
21                    檢 '노조 와해 의혹' 강경훈 삼성전자 부사장 구속영장 청구   
22                   "사회가 건강해야 기업도 발전" 20년 넘은 삼성 나눔의 비결   
23                4차 산업 혁신기업 인탑스, 스마트팩토리 중소벤처기업부 장관상 수상   
24                      '에버랜드 노조와해' 삼성 노사총괄 부사장 내일 영장심사   
25                      文 "수소차에 획기적 뒷받침…비장한 각오로 제조업 부흥"   
26                            코스피, 외인·개인 매도에 2060선으로 밀려   
27                  삼성전자, 美 연말 초대형 TV 판매 웃었다…中·日 추격 따돌려   
28    [부자들은 경제신문을 읽는다] 김성훈 팀장 "테마주, 고점 매수No… 손절가 확인 ...   
29        [돈 잘버는 증권맨에게 듣는다] 이재광 이사 "아난티 이상 급등… 과감히 매수?"   
...                                                 ...   
8250                 [이 종목은 왜?] 日 시장을 노려라! 신고가 행진 '카카오'   
8251                   [베스트리포트]2017 IFA, 테크 르네상스는 '진행중'   
8252                         제이티, 삼성전자와 31억원 규모 공급계약 체결   
8253                     [오늘의포인트]웰 컴백 IT…외인 매수로 주도주 컴백?   
8254                        갤노트8 '흥행몰이'..사전예약 첫날 39.5만대   
8255                        박근혜 前대통령, 구속 만기로 석방? 추가 구속?   
8256                         코스피 약보합…삼성전자·SK하이닉스 1%대 상승   
8257             [장용혁의 마켓시그널] 마지노선에서 반등한 국내증시, 상승세 얼마나?   
8258                                         호텔에서 추석 나기   
8259           [스팟]삼성전자, 2만9000원(1.21%) 오른 243만5000원 출발   
8260                 'TV가 꺼지면 예술로' 삼성전자, 대만서 '더 프레임' 출시   
8261                   삼성전자, 3Q 반도체 호황으로 감익 우려 해소-한화투자證   
8262                             8일 코스피 시초가 약보합 예상-동시호가   
8263                            투자자라면 꼭 봐야할 10대 투자뉴스-8일   
8264             삼성전자, "영업익 3Q 14.5조·4Q 15.5조 낼 것" - 흥국   
8265                                   흔적없는 살인, 증거없는 청탁   
8266                        외환위기 20년, 재계 사라진 그룹과 떠오른 그룹   
8267                            고용 한파 속 채용 늘리는 기업, 이유있네   
8268                       5대 증권사 순익, 전체 절반 육박…양극화 빨라진다   
8269                                머니투데이 오프라인 헤드라인-8일자   
8270                    박근혜, 구속 만기 후 선고 가능성↑…추가 영장 발부될까   
8271   [이 종목은 왜?] '날아 오를' 가능성, 두산인프라코어 vs '악화일로' 에이블씨엔씨   
8272                             스마트폰 '출고가 100만원 시대' 왜?   
8273                 '깜짝증언' 정유라와 갈라선 이경재 변호사 "신뢰관계 깨졌다"   
8274                         '고점부담+北도발'에 증시조정, 배당주는 어때요   
8275                          최순실 측 이경재 변호사, 정유라 변호인 사임   
8276                          신기술 출시 막는 분류체계·제품규정 대폭 손질   
8277           2017 DJSI 발표...삼성전자 1년 만에 재편입 등 총 23곳 편입   
8278                        '기다렸다' 갤노트8 사전예약 첫 날 '흥행예고'   
8279            [시황종합]코스피, 외인·기관 매수 2340대 회복... 6일만의 상승   

                                                   BODY  year month  time  
0     [theL]삼성에버랜드 노조 설립 방해·노조 와해 혐의에버랜드 노조 와해 의혹을 수...  2018    12  1819  
1     [the L]'삼성 노조 와해 의혹'을 수사 중인 검찰이 강경훈 삼성전자 부사장에 ...  2018    12  1819  
2     [특집]삼성전자는 1995년 국내 기업 최초로 사회봉사단을 창단 사회문제 해결을 위...  2018    12  1818  
3     인탑스(주)(대표 김근하 정사진)가 스마트팩토리 도입을 추진해 경북창조경제혁신센터가...  2018    12  1817  
4     염호석씨 

# 시간정제 끝!!

In [12]:
data = twitter.nouns(u'그렇게 기대했던 취업이었지만, 막상 나에게 펼쳐진건 본 게임이었다.')
data

['취업', '막상', '나', '게임']

In [71]:
df1.apply(lambda x : twitter.nouns(x))

AttributeError: 'str' object has no attribute 'apply'

In [66]:
df['TITLE'].apply(lambda x : twitter.nouns(x))

0                             [삼성, 전자, 강경, 훈, 부사, 구속, 심사]
1                   [노조, 의혹, 강경, 훈, 삼성, 전자, 부사, 구속영장, 청구]
2                                [사회, 기업, 발전, 삼성, 나눔, 비결]
3       [차, 산업, 혁신, 기업, 인탑스, 스마트, 팩토리, 중소, 벤처기업, 부, 장관...
4               [에버랜드, 노조, 해, 삼성, 노사, 총괄, 부사, 내일, 영장, 심사]
5                             [수소차, 획기, 뒷받침, 각오, 제조업, 부흥]
6                                    [코스피, 외인, 개인, 매도, 선]
7                               [삼성, 전자, 연말, 초대형, 판매, 추격]
8         [부자, 신문, 김성훈, 팀, 테마주, 점, 매수, 손절, 확인, 후, 매수, 전략]
9       [돈, 잘버, 증권, 맨, 이재, 광, 이사, 아난, 티, 이상, 급등, 과, 감히...
10                         [제조업, 혁신, 노동자, 기업, 양보, 상생, 모델]
11                                [이스탄불, 공항, 장식, 삼성, 사이니]
12                                 [방, 안, 옷가게, 삼성, 패션, 집]
13                         [특징, 주, 삼성, 전자, 사흘, 주, 신저, 경신]
14                                 [뉴욕증시, 하락, 코스피, 선, 약세]
15                                [리포트, 화웨이, 왜, 공공, 적, 이]
16                [부고, 원빈, 머니투데이, 국제, 부, 디자인, 기자, 부친상, 외]
17            

In [ ]:
df['name'] = df['name'].apply(lambda e: e.split()[0])

>>> from timeit import Timer
>>> t = {}
>>> t['methodcaller'] = Timer("map(methodcaller('split', ' '), a)", "from operator import methodcaller; a=['2011-12-22 46:31:11','2011-12-20 20:19:17', '2011-12-20 01:09:21']")
>>> t['lambda'] = Timer("map(lambda s: s.split(), a)", "a = ['2011-12-22 46:31:11','2011-12-20 20:19:17', '2011-12-20 01:09:21']")
>>> t['listcomp'] = Timer("[s.split() for s in a]", "a = ['2011-12-22 46:31:11','2011-12-20 20:19:17', '2011-12-20 01:09:21']")
>>> for name, timer in t.items():
...     print '%s: %.2f usec/pass' % (name, 1000000 * timer.timeit(number=100000)/100000)
... 
listcomp: 2.08 usec/pass
methodcaller: 2.87 usec/pass
lambda: 3.10 usec/pass